mobility/choreographer/src/validator.cc

```diff
@@ -22,17 +22,150 @@
+      ROS_ERROR("Zero keepin zones!! Plan failed");
+      return false;
+    }
+
+    // Based on zones boundries and obstacle map resolution, specify dimensions
+    min = Vec3f(std::floor(min(0) / map_res_) * map_res_,
+                std::floor(min(1) / map_res_) * map_res_,
+                std::floor(min(2) / map_res_) * map_res_) - Vec3f::Ones() * map_res_ * 2.0;
+    max = Vec3f(std::ceil(max(0) / map_res_) * map_res_,
+                std::ceil(max(1) / map_res_) * map_res_,
+                std::ceil(max(2) / map_res_) * map_res_) + Vec3f::Ones() * map_res_ * 2.0;
+
+
+    Vec3f origin = min;
+    Vec3f dimf = (max - min) / map_res_;
+    Vec3i dim(std::ceil(dimf(0)), std::ceil(dimf(1)), std::ceil(dimf(2)));
+    int num_cell = dim(0) * dim(1) * dim(2);
+
+
+    // Declare voxel map
+    std::vector<signed char> map(num_cell, val_unknown_);
+    jps_map_util_.reset(new JPS::VoxelMapUtil());
+    jps_map_util_->setMap(origin, dim, map, map_res_);
+
+    // Keepin/Keepout zones:
+    // To reduce computational load, only the contourn of the keepin/keepout
+    // zones is defined as occupied, to minimize the number of points that
+    // are inflated
+    // 0) The voxel map starts with all voxels set to unknown
+    // 1) Keepin Zones add contour as occupied
+    for (auto &zone : zones_.zones) {
+      if (zone.type == ff_msgs::Zone::KEEPIN) {
+        zmin << std::min(zone.min.x, zone.max.x),
+            std::min(zone.min.y, zone.max.y), std::min(zone.min.z, zone.max.z);
+        zmax << std::max(zone.min.x, zone.max.x),
+            std::max(zone.min.y, zone.max.y), std::max(zone.min.z, zone.max.z);
+        Vec3f tmp = Vec3f::Zero();
+        for (int i = 0; i < 3; i++) {
+          int j = (i + 1) % 3;
+          int k = (i + 2) % 3;
+          for (auto zx = zmin(j); zx <= zmax(j); zx += map_res_) {
+            for (auto zy = zmin(k); zy <= zmax(k); zy += map_res_) {
+              tmp(j) = zx;
+              tmp(k) = zy;
+              tmp(i) = zmin(i) - map_res_ * 1.001;
+              map[jps_map_util_->getIndex(jps_map_util_->floatToInt(tmp))] =
+                  val_occ_;
+              tmp(i) = zmax(i) + map_res_ * 1.001;
+              map[jps_map_util_->getIndex(jps_map_util_->floatToInt(tmp))] =
+                  val_occ_;
+            }
+          }
+        }
+      }
+    }
+    // 2) We set the interior of the keepin zones to free, this corrects the
+    // case where keepin zones are connected and a contourn was put between them
+    for (auto &zone : zones_.zones) {
+      if (zone.type == ff_msgs::Zone::KEEPIN) {
+        zmin << std::min(zone.min.x, zone.max.x),
+            std::min(zone.min.y, zone.max.y), std::min(zone.min.z, zone.max.z);
+        zmax << std::max(zone.min.x, zone.max.x),
+            std::max(zone.min.y, zone.max.y), std::max(zone.min.z, zone.max.z);
+        // add points on surface
+        Vec3f tmp = Vec3f::Zero();
+        for (auto zx = zmin(0); zx <= zmax(0); zx += map_res_) {
```

Comment:
if it's fast enough it doesn't matter, but since there are probably fewer keepin zones than points in the keepin zones, might be faster to check the points as you add them instead

Reply:
I'm not sure I understand, we only iterate through the points belonging to the zone

Reply:
My understanding is, right now you have:

1) Set all to unoccupied
2) Set borders to occupied
3) Set keepins to unoccupied
4) Set keepouts to occupied

I suspect it might be faster to instead do:

1) Set all to unoccupied
2) Set borders to occupied, while looping through keepin zones to check they are not in one
3) Set keepouts to occupied

Since the number of pixels inside all zones is probably bigger than the number of border pixels times the number of zones.

But if it's already fast enough no need to worry about it.

Reply:
But in the second approach we would still have to fill in the keepin zones as free

Reply:
Oh, I thought the map started with everything set to free.

Reply:
I made it start as val_unknown_, so we are able to throw the error VIOLATES_KEEP_IN

Reply:
Ok sounds good!